impala/_rpc/beeswax.py

```diff
@@ -12,17 +12,58 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+from __future__ import absolute_import, print_function
+
+import os
+import sys
+import six
+from six.moves import map
+from six.moves import range
+
 from impala.error import RPCError, QueryStateError, DisconnectedError
-from impala._thrift_gen.beeswax import BeeswaxService
-from impala._thrift_gen.ImpalaService import ImpalaService
-from impala._thrift_gen.Status.ttypes import TStatus, TStatusCode
-from impala._thrift_gen.ExecStats.ttypes import TExecStats
 
-from thrift.transport.TSocket import TSocket
-from thrift.transport.TTransport import TBufferedTransport, TTransportException
-from thrift.protocol.TBinaryProtocol import (
-    TBinaryProtocolAccelerated as TBinaryProtocol)
-from thrift.Thrift import TApplicationException
+if six.PY2:
+    from thrift.transport.TSocket import TSocket
+    from thrift.transport.TTransport import TBufferedTransport, TTransportException
+    from thrift.protocol.TBinaryProtocol import (
+        TBinaryProtocolAccelerated as TBinaryProtocol)
+    from thrift.Thrift import TApplicationException
+
+    from impala._thrift_gen.beeswax import BeeswaxService
+    from impala._thrift_gen.ImpalaService import ImpalaService
+    from impala._thrift_gen.Status.ttypes import TStatus, TStatusCode
+    from impala._thrift_gen.ExecStats.ttypes import TExecStats
+elif six.PY3:
+        # import thriftpy code
+    from thriftpy import load
+    from thriftpy.thrift import TClient, TApplicationException
+    # TODO: reenable cython
+    # from thriftpy.protocol import TBinaryProtocol
+    from thriftpy.protocol.binary import TBinaryProtocol
+    from thriftpy.transport import TSocket, TTransportException
+    # TODO: reenable cython
+    # from thriftpy.transport import TBufferedTransport
+    from thriftpy.transport.buffered import TBufferedTransport
+
+    # dynamically load the thrift modules
+    thrift_dir = os.path.join(os.path.dirname(os.path.dirname(__file__)),
+                              'thrift')
+    ExecStats = load(os.path.join(thrift_dir, 'ExecStats.thrift'),
+                     include_dirs=[thrift_dir])
+    Status = load(os.path.join(thrift_dir, 'Status.thrift'),
+                  include_dirs=[thrift_dir])
+    ImpalaService = load(os.path.join(thrift_dir, 'ImpalaService.thrift'),
+                         include_dirs=[thrift_dir])
+    beeswax = load(os.path.join(thrift_dir, 'beeswax.thrift'),
+                   include_dirs=[thrift_dir])
+    sys.modules[ExecStats.__name__] = ExecStats
+    sys.modules[Status.__name__] = Status
+    sys.modules[ImpalaService.__name__] = ImpalaService
+    sys.modules[beeswax.__name__] = beeswax
+    from ExecStats import TExecStats
+    from Status import TStatus, TStatusCode
+    from ImpalaService import ImpalaService
+    import beeswax as BeeswaxService
```

Comment:
yikes, is there no canonical way to do this? perhaps it would be better to factor this into a thrift_api compatibility module so that you don't have to rely on such black magic